pkg/engine/variables/operator/in.go

```diff
@@ -103,6 +103,74 @@ func keyExistsInArray(key string, value interface{}, log logr.Logger) (invalidTy
 	return false, false
 }
 
+func (in InHandler) validateValueWithStringSetPattern(key []string, value interface{}) (keyExists bool) {
+	invalidType, keyExists := setExistsInArray(key, value, in.log)
+	if invalidType {
+		in.log.Info("expected type []string", "value", value, "type", fmt.Sprintf("%T", value))
+		return false
+	}
+
+	return keyExists
+}
+
+// setExistsInArray checks if the key is a subset of value
+// The value can be a string, an array of strings, or a JSON format
+// array of strings (e.g. ["val1", "val2", "val3"].
+func setExistsInArray(key []string, value interface{}, log logr.Logger) (invalidType bool, keyExists bool) {
+	switch valuesAvailable := value.(type) {
+
+	case []interface{}:
+		var valueSlice []string
+		for _, val := range valuesAvailable {
+			v, ok := val.(string)
+			if !ok {
+				return true, false
+			}
+			valueSlice = append(valueSlice, v)
+		}
+
+		return false, checkSubset(key, valueSlice)
+
+	case string:
+
+		if len(key) == 1 && key[0] == valuesAvailable {
+			return false, true
+		}
+
+		var arr []string
+		if err := json.Unmarshal([]byte(valuesAvailable), &arr); err != nil {
+			log.Error(err, "failed to unmarshal value to JSON string array", "key", key, "value", value)
+			return true, false
+		}
+
+		return false, checkSubset(key, arr)
+
+	default:
+		return true, false
+	}
+}
+
+func checkSubset(key []string, value []string) bool {
+	set := make(map[string]int)
+
+	for _, val := range value {
+		set[val]++
+	}
+
+	for _, val := range key {
+		count, found := set[val]
+		if !found {
+			return false
```

Comment:
Seems that `checkSubset` returns true if **all** keys are found in the value, while for `NotIn` in [this](https://github.com/kyverno/kyverno/issues/1367#issuecomment-741976456) use case, it expects "any of the three matches; not all three together". 

We should define what's the expected behavior for operators `In`/`NotIn`. 

cc @chipzoller @JimBugwadia

Reply:
Given 2 sets, S1 and S2:

* `S1 In S2`: means all values of S1 are in S2
* `S1 NotIn S2`: means none of the values of S1 are in S2

Reply:
Will write two separate functions for In and NotIn` based on this definition then.